File: fetch_images.js
const puppeteer = require('puppeteer-extra');
const StealthPlugin = require('puppeteer-extra-plugin-stealth');
puppeteer.use(StealthPlugin());

const fs = require('fs');
const csv = require('csv-parser');
const fastcsv = require('fast-csv');
const path = require('path');
const https = require('https');

const inputCsv = 'RecipesImp.csv';
const outputCsv = 'Recipes_with_Images_Final.csv';
const imagesDir = 'images';

if (!fs.existsSync(imagesDir)) fs.mkdirSync(imagesDir);
const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));

const cleanFilename = (title) => {
    let clean = title.replace(/\r?\n|\r/g, ' ');
    clean = clean.replace(/[^a-zA-Z0-9\s-]/g, '');
    clean = clean.replace(/\s+/g, '_').replace(/^_+|_+$/g, '');
    return clean;
};

const downloadImage = (url, filepath) => {
    return new Promise((resolve) => {
        try {
            if (url.startsWith('data:image')) {
                const matches = url.match(/^data:image\/([A-Za-z-+\/]+);base64,(.+)$/);
                if (!matches || matches.length !== 3) {
                    resolve(false);
                    return;
                }
                const buffer = Buffer.from(matches[2], 'base64');
                fs.writeFileSync(filepath, buffer);
                resolve(true);
            } else {
                const options = { headers: { 'User-Agent': 'Mozilla/5.0' } };
                https.get(url, options, (res) => {
                    if (res.statusCode === 200) {
                        const fileStream = fs.createWriteStream(filepath);
                        res.pipe(fileStream);
                        fileStream.on('finish', () => { fileStream.close(); resolve(true); });
                    } else { resolve(false); }
                }).on('error', () => resolve(false));
            }
        } catch (e) { resolve(false); }
    });
};

const fetchGoogleImage = async (page, query) => {
    try {
        await page.goto(`https://duckduckgo.com/?q=${encodeURIComponent(query)}&t=h_&iax=images&ia=images`, {
            waitUntil: 'networkidle2',
            timeout: 30000
        });

        await delay(2000);

        const imageUrl = await page.evaluate(() => {
            const imgs = Array.from(document.querySelectorAll('img.tile--img__img'));
            const results = imgs.filter(img => img.src && img.src.length > 10);
            return results.length > 0 ? results[0].src : null;
        });

        return imageUrl;
    } catch (e) {
        console.error(`  [Puppeteer Error: ${e.message}]`);
        return null;
    }
};

const processRows = async (results) => {
    console.log(`Loaded ${results.length} rows.`);

    // Try finding a chrome executable path for Windows
    const executablePaths = [
        "C:\\Program Files\\Google\\Chrome\\Application\\chrome.exe",
        "C:\\Program Files (x86)\\Google\\Chrome\\Application\\chrome.exe",
        "C:\\Program Files\\Microsoft\\Edge\\Application\\msedge.exe",
        "C:\\Program Files (x86)\\Microsoft\\Edge\\Application\\msedge.exe"
    ];
    let executablePath = null;
    for (let path of executablePaths) {
        if (fs.existsSync(path)) {
            executablePath = path;
            break;
        }
    }

    if (!executablePath) {
        console.error("No valid Chrome/Edge installation found.");
        process.exit(1);
    }

    console.log(`Using browser executable: ${executablePath}`);

    const browser = await puppeteer.launch({
        headless: "new",
        executablePath: executablePath,
        args: ['--no-sandbox', '--disable-setuid-sandbox', '--disable-infobars', '--window-position=0,0', '--ignore-certifcate-errors', '--ignore-certifcate-errors-spki-list']
    });

    const page = await browser.newPage();
    await page.setViewport({ width: 1280, height: 800 });

    for (let i = 0; i < results.length; i++) {
        const row = results[i];
        const title = row.title;

        let safeTitle = cleanFilename(title);
        if (safeTitle.length > 50) safeTitle = safeTitle.substring(0, 50);

        const imageFilename = `${safeTitle}.jpg`;
        const imageFilepath = path.join(imagesDir, imageFilename);
        const relativePath = `images/${imageFilename}`;

        if (fs.existsSync(imageFilepath)) {
            console.log(`[${i + 1}/${results.length}] ${title.substring(0, 30)}... -> File exists securely`);
            row.image_path = relativePath;
            continue;
        }

        let cleanQuery = title.replace(/\r?\n|\r/g, ' ');
        cleanQuery = cleanQuery.replace(/\(.*\)/g, '').trim();

        console.log(`[${i + 1}/${results.length}] Fetching for: ${cleanQuery}`);

        let query = `${cleanQuery} recipe food`;
        let imgUrl = await fetchGoogleImage(page, query);

        if (!imgUrl) {
            console.log(`  -> Main query failed, trying fallback: ${cleanQuery}`);
            imgUrl = await fetchGoogleImage(page, `${cleanQuery}`);
        }

        if (imgUrl) {
            let success = await downloadImage(imgUrl, imageFilepath);
            if (success) {
                console.log(`  -> Saved: ${imageFilename}`);
                row.image_path = relativePath;
            } else if (imgUrl.startsWith("http")) { // DuckDuckGo passes real imgur links. Attempt curl if node request blocks it.
                row.image_path = "";
                console.log(`  -> Failed saving via stream. Got URL: ${imgUrl}`);
            } else {
                row.image_path = "";
            }
        } else {
            console.log(`  -> No image found.`);
            row.image_path = "";
        }

        await delay(2000);
    }

    await browser.close();

    fastcsv.writeToPath(outputCsv, results, { headers: true })
        .on('finish', () => {
            console.log(`Done! Saved final dataset to ${outputCsv}`);
        });
};

const main = () => {
    const results = [];
    console.log(`Reading ${inputCsv}...`);

    fs.createReadStream(inputCsv)
        .pipe(csv())
        .on('data', (data) => results.push(data))
        .on('end', () => {
            processRows(results);
        });
};

main();

File: fetch_images.py
import pandas as pd
import requests
import os
import time
from ddgs import DDGS
import re
import warnings

warnings.filterwarnings('ignore')

input_csv = "RecipesImp.csv"
output_csv = "Recipes_with_Images.csv"
images_dir = "images"

os.makedirs(images_dir, exist_ok=True)
print(f"Reading {input_csv}...")
df = pd.read_csv(input_csv)

def clean_filename(title):
    # Remove chars that aren't alphanumeric or spaces/hyphens
    # Also replace newlines with space
    title_str = str(title).replace('\n', ' ').replace('\r', ' ')
    clean = re.sub(r'[^a-zA-Z0-9\s\-]', '', title_str)
    # squish spaces and replace with underscore
    clean = re.sub(r'\s+', '_', clean).strip('_')
    return clean

def fetch_image_url(query):
    for attempt in range(4):
        try:
            results = DDGS().images(query, max_results=1)
            if results and len(results) > 0:
                return results[0]['image']
            break # if no results but no error, break
        except Exception as e:
            err_msg = str(e)
            print(f"  [DDG Error: {err_msg[:50]}...]")
            if "403" in err_msg or "Ratelimit" in err_msg:
                sleep_time = 10 * (attempt + 1)
                print(f"  [Rate limited. Sleeping {sleep_time}s...]")
                time.sleep(sleep_time)
            else:
                break
    return None

def download_image(url, filepath):
    try:
        headers = {'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36'}
        response = requests.get(url, headers=headers, stream=True, timeout=10)
        if response.status_code == 200:
            with open(filepath, 'wb') as f:
                for chunk in response.iter_content(1024):
                    f.write(chunk)
            return True
        else:
            print(f"  [Fetch failed: {response.status_code}]")
    except Exception as e:
        print(f"  [Download exception: {str(e)[:50]}]")
    return False

image_paths = []

print(f"Starting to process {len(df)} recipes...")
for index, row in df.iterrows():
    title = row['title']
    safe_title = clean_filename(title)
    
    # Keep recipe name under reasonable length
    if len(safe_title) > 50:
        safe_title = safe_title[:50]
        
    image_filename = f"{safe_title}.jpg"
    image_filepath = os.path.join(images_dir, image_filename)
    relative_path = f"images/{image_filename}"
    
    # For query, remove weird newlines and parentheses
    clean_query = str(title).replace('\n', ' ').replace('\r', ' ')
    clean_query = re.sub(r'\(.*?\)', '', clean_query).strip()
    
    print(f"[{index+1}/{len(df)}] {clean_query}")
    
    if os.path.exists(image_filepath):
        print(f"  -> Exists: {image_filename}")
        image_paths.append(relative_path)
        continue

    query = f"{clean_query} recipe image"
    image_url = fetch_image_url(query)
    
    if image_url:
        success = download_image(image_url, image_filepath)
        if success:
            print(f"  -> Saved: {image_filename}")
            image_paths.append(relative_path)
            time.sleep(3)
        else:
            print(f"  -> Failed to download")
            image_paths.append("")
            time.sleep(2)
    else:
        print(f"  -> No image found.")
        image_paths.append("")
        time.sleep(3) 

df['image_path'] = image_paths
df.to_csv(output_csv, index=False)
print(f"Done! Saved combined dataset to {output_csv}")
